File: T-Bank/Algorithms-and-Data-Structures-2025/9sprint-SegmentTree/contest/A-Sage/solution.py
import sys


class SegmentTree:
    def __init__(self, n: int) -> None:
        self.n: int = n
        self.interval: list[int] = [0] * (3 * 100000)
    
    def build(self, sequence: list[int]) -> None:
        for i in range(self.n):
            self.interval[self.n + i] = sequence[i]
        for i in range(self.n - 1, 0, -1):
            self.interval[i] = self.interval[i << 1] + self.interval[i << 1 | 1]
    
    def update(self, idx: int, val: int) -> None:
        idx += self.n
        self.interval[idx] = val
        while idx > 1:
            self.interval[idx >> 1] = self.interval[idx] + self.interval[idx ^ 1]
            idx >>= 1
    
    def range_sum(self, left: int, right: int) -> int:
        output: int = 0
        left += self.n 
        right += self.n
        while left < right:
            if left & 1:
                output += self.interval[left]
                left += 1
            if right & 1:
                right -= 1
                output += self.interval[right]
            left >>= 1
            right >>= 1 
        return output
            


def solution() -> None:
    """
    Time Complexity O(NlogN)
    Memory Complexity O(N)
    """
    n, m = map(int, sys.stdin.readline().rstrip().split())
    st: SegmentTree = SegmentTree(n)
    sequence: list[int] = list(map(int, sys.stdin.readline().rstrip().split()))
    st.build(sequence)
    for _ in range(m):
        op, x, y = map(int, sys.stdin.readline().rstrip().split())
        if op == 1: st.update(x, y)
        else: sys.stdout.write('{}\n'.format(st.range_sum(x, y)))


if __name__ == '__main__':
    solution()

File: T-Bank/Algorithms-and-Data-Structures-2025/9sprint-SegmentTree/contest/A-Sage/solution.cpp
#include <iostream>
#include <vector>


class SegmentTree {
private:
    std::vector<long long> interval;
    int n;
public:
    SegmentTree(int n_) : n(n_), interval(3 * 100000, 0) {};

    void build(std::vector<int>& sequence) {
        for (int i = 0; i < sequence.size(); ++i) {
            interval[n + i] = static_cast<long long>(sequence[i]);
        }
        for (int i = n - 1; i > 0; --i) {
            interval[i] = interval[i << 1] + interval[i << 1 | 1];
        }
    }
    
    void update(int idx, int val) {
        interval[n + idx] = val;
        for (int i = idx + n; i > 1; i >>= 1) {
            interval[i >> 1] = interval[i] + interval[i ^ 1];
        }
    }
    
    long long getSum(int left, int right) {
        long long output = 0;
        for (int l = left + n, r = right + n; l < r; l >>= 1, r >>= 1) {
            if (l & 1) output += interval[l++];
            if (r & 1) output += interval[--r];
        }
        return output;
    }
};


void solution() {
    /*
    Time Complexity O(NlogN)
    Memory Complexity O(N)
    */
    int n, m;
    std::scanf("%d %d", &n, &m);
    SegmentTree st(n);
    std::vector<int> sequence(n, 0);
    for (int i = 0; i < n; ++i) std::scanf("%d", &sequence[i]);
    st.build(sequence);
    for (int i = 0; i < m; ++i) {
        int op, x, y;
        std::scanf("%d %d %d", &op, &x, &y);
        if (op == 1) st.update(x, y);
        else std::printf("%lld\n", st.getSum(x, y));
    }
}


int main() {
    solution();
}